src/GraphQL.SystemTextJson/ExecutionResultJsonConverter.cs

```diff
@@ -0,0 +1,138 @@
+using System;
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace GraphQL.SystemTextJson
+{
+    public class ExecutionResultJsonConverter : JsonConverter<ExecutionResult>
+    {
+        public override void Write(Utf8JsonWriter writer, ExecutionResult value, JsonSerializerOptions options)
+        {
+            writer.WriteStartObject();
+
+            // Important: Don't pass in options when recursively calling Serialize.
+            // See docs: https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to
+            WriteData(writer, value);
+            WriteErrors(writer, value.Errors, value.ExposeExceptions);
+            WriteExtensions(writer, value);
+
+            writer.WriteEndObject();
+        }
+
+        private void WriteData(Utf8JsonWriter writer, ExecutionResult result)
+        {
+            var data = result.Data;
+
+            if (result.Errors?.Count > 0 && data == null)
+            {
+                return;
+            }
+
+            writer.WritePropertyName("data");
+            JsonSerializer.Serialize(writer, data);
```

Comment:
Have a test when `data` is `null` and `result.Errors.Count == 0`?

Reply:
Added. See `Writes_Correct_Execution_Result_With_Null_Data_And_Some_Errors`. 

I'm not 100% if graphql-dotnet is 100% to spec on all this. The writers might have a hard time distinguishing between errors occurring before or after execution started, as per spec [here](https://github.com/graphql/graphql-spec/blob/master/spec/Section%207%20--%20Response.md#response-format), but not really tempted to dig too deep into all that. Just solidified some tests around what currently is the behaviour.

Reply:
I also thought about it when I was engaged in a graphql client. We will see.